# 03 Pandas Basics

**Phase:** Phase1_Python_Pandas  
**Status:** 🔴 initialized    
**Date Started:**  

---


First lookhead(), tail()

Structure  shape, ndim, columns, dtypes, info()

Statistics  describe()

Data qualityisna().sum(), duplicated().sum()

In [4]:
import pandas as pd 

In [10]:
data = pd.read_csv(r'../datasets/data_files/Titanic-Dataset.csv')

Load it into a DataFrame

Run your full EDA checklist

Answer these three questions from the data:

How many missing values are in each column?

What are the data types of each column?

How many duplicate rows exist?

In [19]:
# First look on data 
data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [20]:
data.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [25]:
# data strucuture 
(rows, cols), cells, dimension  = data.shape, data.size, data.ndim

rows, cols, cells, dimension

(891, 12, 10692, 2)

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [48]:
# data quality check 
data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [47]:
data.duplicated().sum()

np.int64(0)

In [49]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [58]:
data['Age'].isna().sum() / data.shape[0]

np.float64(0.19865319865319866)

In [59]:
data.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [68]:
data['Fare'].duplicated().sum()

np.int64(643)

This is where it gets interesting. Looking at your output — three quick questions:
1. Cabin has 687 missing out of 891 rows — that's ~77% missing. What would you do with this column and why?
2. Age has 177 missing — would you drop those rows or fill the missing values? And if fill, with what?
3. Embarked has only 2 missing — what's the easiest fix here?

response : 

---

the very first thing as a person related to data have to do is consider recollection of data , there can be any error during loading the data but since this not in question and considering only one data source present i will conder droping this analysis when the missing  values are this vast but again we  are learning and as from that prespective what i can do is jus drop the columns and done , why is that 
to handle null values we have options like 
fill it or not 
if fill then we can use mean, median, mode  and also use grouping as per relatability of data 
but filling the null come at huge risk, cause it can  distort the data and its statistics due to noise and note the filled data is not real world . 
so we can only consider filling the values when the missing values weight less or have less impact well at that time we can just drop again lol , second and main reason to still fill the values is importance of columns 
now cabin is likely have less affect on survival rate, its not like cabins had any life saving things or capsules like in submarine i consider it , so i will drop it rather than filling it 



now for age part i calculated that the missing values are around 19.8 % again the portion is large enough to affect the data, but one thing in emergency aspects like age, gender matters so from one point it can be important factor to predict the survival 
now if we try to fill the missing values , what we can consider is filling them but with consideration cause i notice in describe() part that the age min is around .4 and max is 80 but the mean is 29, considering that mean could be 40, as per min and max means majority could be youngesters and the sd of 14 also tells that on average the difference between mean and age is around 14 , which is huge 
so going with safest choice that is median 
now if we were to consider any relative column i think we can consider pclass we any passenger is in same class it can be based on age which is quite common in real word 

hmm embarked , 2 missing , we can consider fill them based on mode , we can consider fare to be base to group them but fare have over 200 unique so just fill it with mode , since it is string col

conclusion : 

---

Cabin — correct decision and correct reasoning. Dropping at 77% missing is standard practice. And your point about recollection of data first is exactly what a senior analyst would say — check the source before assuming the data is just bad.

Age — this is where you really stood out:

Calculated the missing % mentally — 19.8% is correct ✅

Used describe() output to reason about distribution — that's real analytical thinking ✅

Chose median over mean because of skew risk — correct ✅

Suggested grouping by Pclass for fill — this is called conditional imputation and it's an advanced technique most entry level candidates don't know ✅

Embarked — mode fill for a categorical column with 2 missing values — perfect, exactly right ✅

In [69]:
data['Age'] = data.groupby('Pclass')['Age'].transform(
    lambda x: x.fillna(x.median())
)

In [71]:
data = data.drop(columns='Cabin')

In [79]:
data['Embarked'] = data['Embarked'].fillna(data['Embarked'].mode()[0])

In [81]:
data.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [82]:
print('basic check and quality improvement done')

basic check and quality improvement done
